src/librustc_typeck/check/mod.rs

```diff
@@ -5465,9 +5465,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             .unwrap_or(false);
 
         let (res, self_ctor_substs) = if let Res::SelfCtor(impl_def_id) = res {
-            let ty = self.impl_self_ty(span, impl_def_id).ty;
```

Comment:
Okay, I see, the `.ty` should've been a big warning flag (the other field is... `substs`).

Can you also change the call sites in `check::method::suggest` to just use `tcx.type_of(impl_def_id)`?
And then maybe inline `impl_self_ty` into the only remaining use site (which is in `check::method::confirm`?).

If `impl_self_ty` was used more widely in `check::method` I'd suggest renaming it to `instantiate_impl_self_ty` *and* make it private to `check::method`. But it's not, so might as well get rid of it before this happens again.

Reply:
I would prefer not dealing with it in this PR. I've filed https://github.com/rust-lang/rust/issues/69489 for the follow-up.